### PR TITLE
Results object

### PR DESCRIPTION
**Pull request by okeefm**
This Pull Request adds an `IndividualResult` object to Javascript and Ruby, which allows for the storage of results about a particular patient-measure combination. Also includes several changes to the `Measure` model to accommodate the relations to `IndividualResult`. 

Pull requests into cqm-models require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable ("N/A") and :white_check_mark:.

**Submitter:**
- [x] This pull request describes why these changes were made.
- [x] Internal ticket for this PR: https://jira.mitre.org/browse/TACO-48
- [x] Internal ticket links to this PR
- [x] Code diff has been done and been reviewed
- [x] Tests are included and test edge cases N/A
- [x] Tests have been run locally and pass N/A
- [ ] If there were any JavaScript changes, this PR has updated the `dist` directory using `npm run dist`
- [ ] If applicable, the library version number in `package.json` and `cqm-models.gemspec` has been updated

**Reviewer 0:**

Name: @holmesie
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

**Reviewer 1:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

**Reviewer 2:**

Name: @ljades
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

### CODE REVIEW

File: app/models/qdm/tacoma/individual_result.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    field :MSRPOPL, type: Integer
+    field :OBSERV, type: Float
+    field :MSRPOPLEX, type: Integer
+
+    # Result Attributes
+    field :clause_results, type: Hash
+    field :episode_results, type: Hash
+    filed :population_relevance, type: Hash
+    field :statement_relevance, type: Hash
+    field :statement_results, type: String
```
Comment (ljades) [inline app/models/qdm/tacoma/individual_result.rb]: In the Javascript version this is a Mixed, but here it's a String.

Comment (okeefm) [reply app/models/qdm/tacoma/individual_result.rb]: Good catch. This should be a `Mixed`/`Hash`.

File: app/assets/javascripts/IndividualResult.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
+const mongoose = require('mongoose');
+
+const [mNumber, mString, Mixed, ObjectId] = [
```
Comment (holmesie) [inline app/assets/javascripts/IndividualResult.js]: Any particular reason for appending these with `m`?

Comment (okeefm) [reply app/assets/javascripts/IndividualResult.js]: @ljades brought this up as well, so I'll tag him in the explanation: at one point in the development process for some of the other models, I was running into clashes with the Javascript builtins for `Number` `String`, etc., so I appended the `m` to avoid bumping into those namespaces. It's possible those issues were coming from an unrelated source, but when I was working on that, using the `m` fixed them, so I decided to duplicate it here. I can swap them all out for their un-appended versions and see what it breaks.

Comment (holmesie) [reply app/assets/javascripts/IndividualResult.js]: Agreed, I think that is the way to go.

Comment (okeefm) [reply app/assets/javascripts/IndividualResult.js]: Ok, I'll revert these, and we can bring them back in another PR.

File: app/models/qdm/tacoma/measure.rb [hunk 3]
```diff
@@ -72,8 +72,9 @@ class Measure
     # Relations to other model classes
     belongs_to :user
     belongs_to :bundle, class_name: 'HealthDataStandards::CQM::Bundle'
-    has_and_belongs_to_many :records, inverse_of: nil
+    has_and_belongs_to_many :patients, inverse_of: nil
     has_one :package, class_name: 'CqlMeasurePackage', inverse_of: :measure, dependent: :delete
+    has_many :individual_results
```
Comment (jbradl11) [inline app/models/qdm/tacoma/measure.rb]: If there is a relationship of Patient to result, and Patient to measure, do you also need Measure to Results?

File: app/models/qdm/tacoma/individual_result.rb [hunk 4]
```diff
@@ -0,0 +1,33 @@
+module QDM
+  # IndividualResult stores the patient-level (population/clause) results for a patient/measure combination
+  class IndividualResult
```
Comment (jbradl11) [inline app/models/qdm/tacoma/individual_result.rb]: Is this Result object storing information for one population of a measure and there would be many results for many populations?

Comment (okeefm) [reply app/models/qdm/tacoma/individual_result.rb]: This Result object is storing information for an entire measure, but only one patient. My intention with the referenced comment was to note that the clause-level results (e.g. whether the patient satisfies a clause) is included in this Result object.

File: app/models/qdm/tacoma/measure.rb (hunk 3 above)
Comment (okeefm) [reply app/models/qdm/tacoma/measure.rb]: I considered this. I was trying to think of a scenario where getting all the results for a Measure directly would be more useful than being able to get them through the Patients.

Comment (okeefm) [reply app/models/qdm/tacoma/measure.rb]: Even in Bonnie, it's my understanding that Result access goes through the Patients, rather than through Measures.

File: app/assets/javascripts/IndividualResult.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    // Population Attributes
+    STRAT: Number,
+    IPP: Number,
+    DENOM: Number,
+    NUMER: Number,
+    NUMEX: Number,
+    DENEX: Number,
+    DENEXCEP: Number,
+    MSRPOPL: Number,
+    OBSERV: Number,
```
Comment (okeefm) [inline app/assets/javascripts/IndividualResult.js]: This should be an Array. Double check the contents of that array.

File: app/assets/javascripts/IndividualResult.js [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+const [Number, String, Mixed, ObjectId] = [
+  mongoose.Schema.Types.Number,
+  mongoose.Schema.Types.String,
+  mongoose.Schema.Types.Mixed,
+  mongoose.Schema.Types.ObjectId,
+];
+
+const IndividualResultSchema = mongoose.Schema(
+  {
+    // Population Attributes
```
Comment (okeefm) [inline app/assets/javascripts/IndividualResult.js]: Encapsulate these in an object so they can be iterated over, and so the types can be different between episode/population-based measures

Comment (okeefm) [reply app/assets/javascripts/IndividualResult.js]: If episode-based: each population attribute would return an Array of episode results

File: app/assets/javascripts/IndividualResult.js (hunk 5 above)
Comment (okeefm) [reply app/assets/javascripts/IndividualResult.js]: Could be its own Object that would contain the episode result (if there is one), the value, & the units

File: app/assets/javascripts/IndividualResult.js [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    NUMEX: Number,
+    DENEX: Number,
+    DENEXCEP: Number,
+    MSRPOPL: Number,
+    OBSERV: Number,
+    MSRPOPLEX: Number,
+
+    // Result Attributes
+    clause_results: Mixed,
+    episode_results: Mixed,
```
Comment (okeefm) [inline app/assets/javascripts/IndividualResult.js]: Can remove `episode_results` if the above changes to the population attributes are made.

File: app/assets/javascripts/IndividualResult.js [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    population_relevance: Mixed,
+    statement_relevance: Mixed,
```
Comment (okeefm) [inline app/assets/javascripts/IndividualResult.js]: These are used on the Bonnie front-end, and are inferred from other information in the Result object, so can probably be moved out of being stored.

File: app/assets/javascripts/IndividualResult.js [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    statement_results: Mixed,
+
+    // Calculation State attributes
+    state: String,
```
Comment (okeefm) [inline app/assets/javascripts/IndividualResult.js]: Turn this into an enum of possible states

File: app/models/qdm/tacoma/individual_result.rb [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+    field :DENEXCEP, type: Integer
+    field :MSRPOPL, type: Integer
+    field :OBSERV, type: Float
+    field :MSRPOPLEX, type: Integer
+
+    # Result Attributes
+    field :clause_results, type: Hash
+    field :episode_results, type: Hash
+    field :statement_results, type: Hash
+
```
[the remark this answers is not in the document]
Comment (okeefm) [reply app/models/qdm/tacoma/individual_result.rb]: Added as `extendedData` because that's what we called it in the `Patient` model.

**Comment by dczulada** [comment]
templates/index_template.js.erb  and app/assets/javascripts/index.js will also need to be updated to reference Individual Result

module.exports.IndividualResult = require('./IndividualResult').IndividualResult;
module.exports.IndividualResultSchema = require('./IndividualResult').IndividualResultSchema;

File: app/models/qdm/tacoma/individual_result.rb (hunk 4 above)
Comment (edeyoung) [reply app/models/qdm/tacoma/individual_result.rb]: @okeefm please follow-up w @jbradl11 on this. doesn't seem like it's structured to handle multiple pops/strats for a single patient for an entire measure. I'm probably missing the conclusion of this conversation as it's from a while ago.
